Models are not magic.
They are hypotheses tested on data.
Writing hypotheses and reflections made the experiment feel purposeful rather than mechanical.

# Non-Linear Data Experiment

## Research Question
Can linear regression model non-linear data?

## Hypothesis
A linear regression model will perform poorly on non-linear data.


In [1]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures


In [2]:
rng = np.random.default_rng(seed=42)
X = rng.random((200, 1)) * 5
y = X.squeeze()**2 + rng.standard_normal(200) * 2

### Data Description
- y is proportional to x² (quadratic relationship)
- Added noise to simulate real-world data
- Expect linear model to struggle


In [3]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [4]:
#Linear model
lin_model = LinearRegression()
lin_model.fit(X_train, y_train)

y_pred_lin = lin_model.predict(X_test)
mse_lin = mean_squared_error(y_test, y_pred_lin)

mse_lin

8.292388869218822

### Linear Model Result
- High MSE indicates poor fit
- Linear model cannot capture curvature

In [5]:
# Introduce non-linearity (Polinomial features)
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

This code performs feature engineering to transform linear input data into polynomial terms. This allows a standard linear regression model to capture non-linear (curved) relationships in your data. 

poly = PolynomialFeatures(degree=2)This initializes the transformer. A degree=2 setting means it will generate all combinations of features up to the power of 2.
Example: If your input is a single feature ([a]), it generates ([1,a,a^{2}]).
Example: If your input has two features \([a,b]\), it generates \([1,a,b,a^{2},ab,b^{2}]\).

X_train_poly = poly.fit_transform(X_train)This performs two actions on your training data in one step:
fit: The transformer "learns" the number of input features and the required output structure.
transform: It actually creates the new polynomial columns (like squared terms and interactions) and returns the expanded feature matrix.

X_test_poly = poly.transform(X_test)This applies the exact same transformation to your test data.
Crucial Note: You only use transform() here (not fit_transform) to ensure the test data is handled using the parameters learned from the training set. This prevents "data leakage" and ensures your model is tested on features structured exactly like those it was trained on. 

Why use this? Linear models can only draw straight lines. By adding \(X^{2}\) as a new feature, you effectively turn a "Linear Regression" model into a "Polynomial Regression" model, allowing it to fit curves and capture more complex patterns in the data.

In [6]:
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)

y_pred_poly = poly_model.predict(X_test_poly)
mse_poly = mean_squared_error(y_test, y_pred_poly)

mse_poly

4.129033932418055

## Comparison
- Linear MSE: 8.2924
- Polynomial MSE: 4.129

### Interpretation
- Adding non-linearity significantly reduced error
- Model complexity matters when data is non-linear


This is exactly what neural networks do:

Polynomial features = manual non-linearity

Neural networks = learned non-linearity

## Reflection
- Linear models fail on non-linear relationships
- Introducing non-linearity dramatically improves performance
- This motivates neural networks and deep learning
- Model power must match data complexity.
